### ISSUE TITLE
dynamic cpusets

### ISSUE THREAD

**Issue by ompiteam** [open-mpi/hwloc#35]
The dyncpuset branch might be mergeable now. The remaining possible optimization are orthogonal, not required before entering trunk. I'd like some feedback about the current implementation.

FWIW, below is the duration (in microseconds) of hwloc_topology_load() depending on the topology size/hierarchy. It's a synthetic topo, so load() does pretty much nothing apart from allocating objects and manipulating tons of cpusets to insert in the tree and compute the {allowed,complete,online}_{cpuset,nodeset}.

{{{
                                                   size    trunk     dyncpusets
  synthetic proc:4                                 4       100       100
  synthetic proc:32                                32      745       413
  synthetic node:4 die:4 core:4 proc:4             256     7750      5409
  synthetic proc:256                               256     41932     34773
  synthetic mach:4 node:4 die:4 core:4 proc:4      1024    44215     49406
  synthetic proc:1024                              1024    1237049   1442945
  synthetic m:4 n:4 d:4 cache:4 core:4 4           4096    X         700547
  synthetic m:4 n:4 d:4 cache:4 cache:4 core:4 4   16384   X         11597185
}}}

In short, dyncpusets decrease memory waste, and do not increase CPU cycles.

1024 is the current static size in the trunk, that's why it's faster than dyncpusets. The dyncpusets branch works at least until 16384 in the above test but the lstopo time became too long for me :)
